File: lib_picker/src/main/java/cn/cheney/lib_picker/view/CaptureLayer.java
package cn.cheney.lib_picker.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import cn.cheney.lib_picker.R;


public class CaptureLayer extends FrameLayout {

    private LinearLayout doneLayer;

    private View rootLayer;
    private ImageView doneIv;
    private ImageView cancelIv;
    private ImageView redCircleIv;
    private ImageView redSquareIv;
    private CircleProgressView progressView;

    private CaptureListener listener;


    public interface CaptureListener {
        void onClick();

        void onLongClick();

        void onBackClick();

        void onDoneClick();
    }


    public CaptureLayer(@NonNull Context context) {
        this(context, null);
    }

    public CaptureLayer(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CaptureLayer(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView();
    }

    private void initView() {
        View rootView = View.inflate(getContext(), R.layout.xpicker_layter_record_capture, null);
        rootLayer = rootView.findViewById(R.id.common_capture_root);
        progressView = rootView.findViewById(R.id.common_capture_action_iv);
        doneLayer = rootView.findViewById(R.id.common_capture_done_layer);
        doneIv = rootView.findViewById(R.id.common_capture_done_iv);
        cancelIv = rootView.findViewById(R.id.common_capture_cancel_iv);
        redCircleIv = rootView.findViewById(R.id.media_record_red_circle);
        redSquareIv = rootView.findViewById(R.id.media_record_red_square);

        progressView.setOnClickListener(v -> {
            if (null != listener) {
                listener.onClick();
            }
        });
        progressView.setOnLongClickListener(v -> {
            if (null != listener) {
                listener.onLongClick();
            }
            return false;
        });

        doneIv.setOnClickListener(v -> {
            if (null != listener) {
                listener.onDoneClick();
            }
        });
        cancelIv.setOnClickListener(v -> {
            if (null != listener) {
                listener.onBackClick();
            }
        });
        addView(rootView);
        normal();
    }


    public void setMaxLength(int maxLength) {
        progressView.setMaxProgress(maxLength);
    }


    public void setProgress(int progress) {
        progressView.setProgress(progress);
    }


    public void normal() {
        progressView.setVisibility(VISIBLE);
        progressView.setProgress(0);
        doneLayer.setVisibility(GONE);
        redCircleIv.setVisibility(VISIBLE);
        redSquareIv.setVisibility(GONE);
    }


    public void recording() {
        progressView.setVisibility(VISIBLE);
        doneLayer.setVisibility(GONE);
        redCircleIv.setVisibility(GONE);
        redSquareIv.setVisibility(VISIBLE);
    }


    public void done() {
        doneLayer.setVisibility(VISIBLE);

        TranslateAnimation cancelAnimation = new TranslateAnimation(Animation.RELATIVE_TO_SELF,
                1, Animation.RELATIVE_TO_SELF, 0,
                0, 0, 0, 0);
        cancelAnimation.setDuration(200);
        cancelIv.setAnimation(cancelAnimation);
        cancelAnimation.start();

        TranslateAnimation doneAnimation = new TranslateAnimation(Animation.RELATIVE_TO_SELF,
                -1, Animation.RELATIVE_TO_SELF, 0,
                0, 0, 0, 0);
        doneAnimation.setDuration(200);
        doneIv.setAnimation(doneAnimation);
        doneAnimation.start();

        progressView.setVisibility(GONE);
        redCircleIv.setVisibility(GONE);
        redSquareIv.setVisibility(GONE);
    }


    public void setListener(CaptureListener listener) {
        this.listener = listener;
    }
}

File: lib_picker/src/main/java/cn/cheney/lib_picker/view/CaptureLayer.kt
package cn.cheney.lib_picker.view

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import android.util.AttributeSet
import android.view.View
import android.view.View.OnClickListener
import android.view.animation.Animation
import android.view.animation.TranslateAnimation
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import cn.cheney.lib_picker.MIXED
import cn.cheney.lib_picker.ONLY_CAPTURE
import cn.cheney.lib_picker.ONLY_RECORDER
import cn.cheney.lib_picker.R
import cn.cheney.lib_picker.callback.CaptureListener

class CaptureLayer @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private lateinit var doneLayer: LinearLayout
    private lateinit var captureBtnLayer: FrameLayout
    private lateinit var doneIv: ImageView
    private lateinit var cancelIv: ImageView
    private lateinit var captureBtn: CaptureButton
    private lateinit var tipTxt: TextView
    private var listener: CaptureListener? = null

    init {
        initView()
    }

    private fun initView() {
        val rootView =
            View.inflate(context, R.layout.xpicker_layer_record_capture, null)
        captureBtnLayer = rootView.findViewById(R.id.capture_btn_layer)
        doneLayer = rootView.findViewById(R.id.xpicker_capture_done_layer)
        doneIv = rootView.findViewById(R.id.xpicker_capture_done_iv)
        cancelIv = rootView.findViewById(R.id.xpicker_capture_cancel_iv)
        tipTxt = rootView.findViewById(R.id.capture_tip_txt)
        addCaptureBtn()
        doneIv.setOnClickListener(OnClickListener {
            listener?.ok()
        })
        cancelIv.setOnClickListener(OnClickListener { v: View? ->
            listener?.cancel()
        })
        addView(rootView)
        reset()
    }

    @SuppressLint("SetTextI18n")
    private fun addCaptureBtn() {
        val screenWidth = Resources.getSystem().displayMetrics.widthPixels
        captureBtn = CaptureButton(context, (screenWidth / 4.5f).toInt())
        captureBtn.setCaptureLisenter(object : CaptureListener() {
            override fun takePictures() {
                listener?.takePictures()
            }

            override fun recordShort(time: Long) {
                listener?.recordShort(time)
            }

            override fun recordTime(time: Long) {
                listener?.recordTime(time)
                tipTxt.text = "${time / 1000}s / ${captureBtn.duration / 1000}s "
            }

            override fun recordStart() {
                listener?.recordStart()
                doneLayer.visibility = View.GONE
            }

            override fun recordEnd(time: Long) {
                listener?.recordEnd(time)
                tipTxt.text = "${time / 1000}s"
            }

            override fun recordZoom(zoom: Float) {
                listener?.recordZoom(zoom)
            }
        })
        captureBtnLayer.addView(captureBtn)
    }


    fun reset() {
        captureBtn.resetState()
        captureBtn.visibility = View.VISIBLE
        tipTxt.visibility = View.VISIBLE
        setTipByType()
        doneLayer.visibility = View.GONE
    }


    fun done() {
        captureBtn.visibility = View.GONE
        tipTxt.visibility = View.GONE

        doneLayer.visibility = View.VISIBLE
        val cancelAnimation = TranslateAnimation(
            Animation.RELATIVE_TO_SELF,
            1f, Animation.RELATIVE_TO_SELF, 0f,
            0, 0f, 0, 0f
        )
        cancelAnimation.duration = 200
        cancelIv.animation = cancelAnimation
        cancelAnimation.start()
        val doneAnimation = TranslateAnimation(
            Animation.RELATIVE_TO_SELF,
            (-1).toFloat(), Animation.RELATIVE_TO_SELF, 0f,
            0, 0f, 0, 0f
        )
        doneAnimation.duration = 200
        doneIv.animation = doneAnimation
        doneAnimation.start()
    }

    fun setListener(listener: CaptureListener?) {
        this.listener = listener
    }

    fun setMaxDuration(duration: Int) {
        if (duration > 1000) {
            captureBtn.duration = duration
        }

    }

    fun setMinDuration(duration: Int) {
        if (duration > 1000) {
            captureBtn.setMinDuration(duration)
        }
    }


    fun setCameraType(type: String) {
        captureBtn.setButtonFeatures(type)
        setTipByType()
    }


    private fun setTipByType() {
        when (captureBtn.type) {
            MIXED -> {
                tipTxt.text = context.getText(R.string.xpicker_mixed_tip)
            }
            ONLY_CAPTURE -> {
                tipTxt.text = context.getText(R.string.xpicker_capture_tip)
            }
            ONLY_RECORDER -> {
                tipTxt.text = context.getText(R.string.xpicker_recorder_tip)
            }
        }
    }
}
